packages/crypto/src/proof.ts

```diff
@@ -0,0 +1,263 @@
+  try {
+    signatureBytes = uint8ArrayFromHexString(appProof.signature);
+  } catch (error) {
+    throw new Error("Failed to decode signature");
+  }
+
+  if (signatureBytes.length !== 64) {
+    throw new Error(
+      `Expected 64 bytes signature (r||s), got ${signatureBytes.length} bytes`,
+    );
+  }
+
+  // Hash the proof payload
+  const payloadBytes = new TextEncoder().encode(appProof.proofPayload);
+  const payloadDigest = sha256(payloadBytes);
+
+  // Verify ECDSA signature
+  const isValid = p256.verify(signatureBytes, payloadDigest, signingKeyBytes);
+  if (!isValid) {
+    throw new Error("Signature verification failed");
+  }
+}
+
+// Pinned AWS Nitro Enclaves Root
+const AWS_ROOT_CERT_PEM = `-----BEGIN CERTIFICATE-----
+MIICETCCAZagAwIBAgIRAPkxdWgbkK/hHUbMtOTn+FYwCgYIKoZIzj0EAwMwSTEL
+MAkGA1UEBhMCVVMxDzANBgNVBAoMBkFtYXpvbjEMMAoGA1UECwwDQVdTMRswGQYD
+VQQDDBJhd3Mubml0cm8tZW5jbGF2ZXMwHhcNMTkxMDI4MTMyODA1WhcNNDkxMDI4
+MTQyODA1WjBJMQswCQYDVQQGEwJVUzEPMA0GA1UECgwGQW1hem9uMQwwCgYDVQQL
+DANBV1MxGzAZBgNVBAMMEmF3cy5uaXRyby1lbmNsYXZlczB2MBAGByqGSM49AgEG
+BSuBBAAiA2IABPwCVOumCMHzaHDimtqQvkY4MpJzbolL//Zy2YlES1BR5TSksfbb
+48C8WBoyt7F2Bw7eEtaaP+ohG2bnUs990d0JX28TcPQXCEPZ3BABIeTPYwEoCWZE
+h8l5YoQwTcU/9KNCMEAwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUkCW1DdkF
+R+eWw5b6cp3PmanfS5YwDgYDVR0PAQH/BAQDAgGGMAoGCCqGSM49BAMDA2kAMGYC
+MQCjfy+Rocm9Xue4YnwWmNJVA44fA0P5W2OpYow9OYCVRaEevL8uO1XYru5xtMPW
+rfMCMQCi85sWBbJwKKXdS6BptQFuZbT73o/gBh1qUxl/nNr12UO8Yfwr6wPLb+6N
+IwLz3/Y=
+-----END CERTIFICATE-----`;
+
+// Official SHA-256 fingerprint
+const AWS_ROOT_CERT_SHA256 =
+  "641A0321A3E244EFE456463195D606317ED7CDCC3C1756E09893F3C68F79BB5B";
+
+export async function verifyCertificateChain(
+  cabundle: Uint8Array[],
+  rootCertPem: string,
+  leafCert: Uint8Array,
+  timestampMs: number,
+): Promise<void> {
+  try {
+    // Check root and assert fingerprint
+    const rootX509 = new x509.X509Certificate(rootCertPem);
+    const rootDer = Buffer.from(new Uint8Array(rootX509.rawData));
+    const rootSha = createHash("sha256")
+      .update(rootDer)
+      .digest("hex")
+      .toUpperCase();
+    if (rootSha !== AWS_ROOT_CERT_SHA256) {
+      throw new Error(
+        `Pinned AWS root fingerprint mismatch: expected=${AWS_ROOT_CERT_SHA256} actual=${rootSha}`,
+      );
+    }
+
+    // Bundle starts with root certificate. We're replacing the root with our hardcoded known certificate, so remove first element
+    const bundleWithoutRoot = cabundle.slice(1);
+    const intermediatesX509 = bundleWithoutRoot.map((c) => {
+      if (!c) throw new Error("Invalid certificate data in cabundle");
+      return new x509.X509Certificate(c);
+    });
+    const leaf = new x509.X509Certificate(leafCert);
+
+    // Build path leaf → intermediates → root, with our hardcoded known root certificate
+    const builder = new x509.X509ChainBuilder({
+      certificates: [rootX509, ...intermediatesX509],
+    });
+    const path = await builder.build(leaf);
```

Comment:
It looks like build actually does verification:

Build iterates through chain https://github.com/PeculiarVentures/x509/blob/08f3fd86824d5b890fed349f64554cc2f413c721/src/x509_chain_builder.ts#L45

and calls findIssuer

https://github.com/PeculiarVentures/x509/blob/08f3fd86824d5b890fed349f64554cc2f413c721/src/x509_chain_builder.ts#L61

which verifies the cert with signatureOnly
https://github.com/PeculiarVentures/x509/blob/08f3fd86824d5b890fed349f64554cc2f413c721/src/x509_chain_builder.ts#L86